### ISSUE TITLE
Comparing Cytominer and Pycytominer Profiles

### ISSUE THREAD

**Issue by gwaygenomics** [broadinstitute/lincs-cell-painting#28]
In this issue, I will discuss results of step 3 outlined in https://github.com/broadinstitute/lincs-cell-painting/issues/22#issue-610266067

Note that this is copied and pasted from a notebook that will be added in a future pull request. Details in this notebook will guide our discussion of the results

# Comparing Pycytominer and Cytominer Processing

We have previously processed all of the Drug Repurposing Hub Cell Painting Data using [cytominer](https://github.com/cytomining/cytominer). Cytominer is an R based image-based profiling tool. In this repo, we reprocess the data with [pycytominer](https://github.com/cytomining/pycytominer). As the name connotes, pycytominer is a python based image-based profiling tool.

We include all processing scripts and present the pycytominer profiles in this open source repository. The repository represents a unified bioinformatics pipeline applied to all Cell Painting Drug Repurposing Profiles. In this notebook, we compare the resulting output data between the processing pipelines for the two tools: Cytominer and pycytominer.

We output several metrics comparing the two approaches

## Metrics

In all cases, we calculate the element-wise absolute value difference between pycytominer and cytominer profiles.

1. Mean, median, and sum of element-wise differencs
2. Per feature mean, median, and sum of element-wise differences
3. Feature selection procedure differences per feature (level 4b only)

In addition, we confirm alignment of the following metadata columns:

* Well
* Broad Sample Name
* Plate

Other metadata columns are not expected to be aligned. For example, we have [updated MOA and Target information](https://github.com/broadinstitute/lincs-cell-painting/issues/11) in the pycytominer version.

## Data Levels

Image-based profiling results in the following output data levels. We do not compare all data levels in this notebook.

| Data | Level | Comparison |
| :---- | :---- | :-------- |
| Images | Level 1 | NA |
| SQLite File (single cell profiles ) | Level 2 | NA |
| Aggregated Profiles with Well Information (metadata) | Level 3 | Yes  |
| Normalized Aggregated Profiles with Metadata | Level 4a | Yes | 
| Normalized and Feature Selected Aggregated Profiles with Metadata | Level 4b | Yes |
| Perturbation Profiles created Summarizing Replicates  | Level 5 | No |

**Comment by gwaygenomics** [comment]
# Summary

![summary_metrics_full](https://user-images.githubusercontent.com/7340421/81506181-8ae81e00-92c2-11ea-907c-5fe8a15fe910.png)

![summary_metrics_zoom](https://user-images.githubusercontent.com/7340421/81506187-92a7c280-92c2-11ea-8fa7-2ac88ac04459.png)

**Figure Legend:** The top plot above represents the complete summary within each plate. The bottom plot above zooms in to absolute difference < 10. Each point represents the mean/median/sum of absolute value differences comparing cytominer and pycytominer. I derived the metrics using every matrix entry per plate. There is one point per plate.

## Interpretation

It is odd that a handful of plates sit around `mean = 2.5; median = 2` in level 3 data. This level data is only derived from aggregating single cell profiles. In level 4a and level 4b data, the mean and median per plate difference sits around 0.5. In level 4a data, the extreme outliers in mean and sum are driven largely by only a few very large outliers (we don't see this in medians). These features are removed in the level 4b data. Some extreme outliers in the level 4a outliers may be explained by the handful of plates that are oddly different in level 3 data.

# Within Level Summaries

## Level 3 - Aggregated Profiles

![level_3_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81506397-fda5c900-92c3-11ea-8c28-1bdea2ab3078.png)

![level_3_metrics_per_plate](https://user-images.githubusercontent.com/7340421/81506416-2037e200-92c4-11ea-991f-5ecc7af3d994.png)

**Figure Legend:** Each point represents a per feature average within a single plate. The top plot above shows a per feature view, while the bottom plot shows a per plate view. 

## Level 4a - Aggregated and Normalized Profiles

![level_4a_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81506452-67be6e00-92c4-11ea-8173-f612d451ef3e.png)

![level_4a_metrics_per_plate](https://user-images.githubusercontent.com/7340421/81506461-7c026b00-92c4-11ea-8af3-3b4644b8cd98.png)

**Figure Legend:** Each point represents a per feature average within a single plate. The top plot above shows a per feature view, while the bottom plot shows a per plate view. 

## Level 4b - Aggregated and Normalized Profiles

![level_4b_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81506495-b10ebd80-92c4-11ea-90df-acaebec4a939.png)

![level_4b_metrics_per_plate](https://user-images.githubusercontent.com/7340421/81506498-b409ae00-92c4-11ea-8356-e9237bf7bd9b.png)

![level_4b_all_across_well_summary](https://user-images.githubusercontent.com/7340421/81506508-c4ba2400-92c4-11ea-8918-98f267706380.png)

**Figure Legend:** Each point represents a per feature average within a single plate. The top plot above shows a per feature view, while the middle plot shows a per plate view. The bottom plot represents the distribution of feature differences in total. Most differences are very near zero.

**Comment by gwaygenomics** [comment]
## Recommendation

We knew at the onset that aligning the profiles derived from each tool was going to be an arduous task. We knew that some cytominer-derived profiles were normalized differently (https://github.com/broadinstitute/lincs-cell-painting/issues/3#issuecomment-591994451), which also implies that different plates could have been processed at different times with different methods. We do not know the extent to which different plates have been processed differently with cytominer. We do know that all plates have been uniformly processed with the same pycytominer pipeline.

Given the consistency and documentation of the pycytominer pipeline, the cytominer processing differences, and the amount of time required to _completely resolve_ these differences (we'd likely need to uniformly process all plates with a new cytominer pipeline as well), I think we should stick with the pycytominer profiles. This repo can exist under additional development after releasing version 1 profiles. If we find future issues with this data, or update the processing pipeline, there is nothing stopping us from releasing future versions.

**Comment by gwaygenomics** [comment]
The code to generate these results is provided in #29 

cc'ing @niranjchandrasekaran @shntnu and @annecarpenter - any feedback/suggestions/comments are welcome!

**Comment by shntnu** [comment]
> It is odd that a handful of plates sit around `mean = 2.5; median = 2` in level 3 data.

These very likely correspond to the 10 plates listed here https://github.com/broadinstitute/lincs-cell-painting/issues/3#issuecomment-591994451

**Comment by shntnu** [comment]
> In level 4a and level 4b data, the mean and median per plate difference sits around 0.5.

Can you test if its related to m.a.d.

R https://stat.ethz.ch/R-manual/R-devel/library/stats/html/mad.html
Python https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.median_absolute_deviation.html

**Comment by shntnu** [comment]
> We knew that some cytominer-derived profiles were normalized differently ([#3 (comment)](https://github.com/broadinstitute/lincs-cell-painting/issues/3#issuecomment-591994451))

Ah – they were *aggregated* differently, not *normalized* differently

**Comment by shntnu** [comment]
> ## Recommendation

> Given the consistency and documentation of the pycytominer pipeline, the cytominer processing differences, and the amount of time required to _completely resolve_ these differences (we'd likely need to uniformly process all plates with a new cytominer pipeline as well), I think we should stick with the pycytominer profiles. This repo can exist under additional development after releasing version 1 profiles. If we find future issues with this data, or update the processing pipeline, there is nothing stopping us from releasing future versions.

I'm on board with this plan! 💯

**Comment by gwaygenomics** [comment]
> These very likely correspond to the 10 plates listed here #3 (comment).
> When creating the notebook, do leave those plates out.

![level_3_metrics_per_plate_uniformplates](https://user-images.githubusercontent.com/7340421/81594614-3363b400-938f-11ea-9388-bbff78b5647b.png)

![level_3_metrics_per_feature_uniformplates](https://user-images.githubusercontent.com/7340421/81594641-3eb6df80-938f-11ea-9b86-93581f089e88.png)

Confirmed that those are the 10 plates. Some features (not all) still do have a bit of a wobble.

> Ah – they were aggregated differently, not normalized differently

Ah! I got it - makes sense.

>> In level 4a and level 4b data, the mean and median per plate difference sits around 0.5.

> Can you test if it is related to m.a.d.

Digging into this now. Three things to note in the analysis log linked in https://github.com/broadinstitute/lincs-cell-painting/issues/3#issuecomment-591994451.

1. The file shows [normalization by DMSO control](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log.sh#L52). For some reason, I thought that the cytominer profiles were whole-plate normalized. This could be a reason for the discrepancy!!
2. Also, the default [`cytominer_scripts/normalize`](https://github.com/broadinstitute/cytominer_scripts/blob/master/normalize.R#L13) is "robustize" but the default [`cytominer/normalize`](https://github.com/cytomining/cytominer/blob/master/R/normalize.R#L34) is "standardize". 
    * Sidenote: The version of `cytominer/normalize` used in processing the cytominer profiles was likely the [August 15, 2016](https://github.com/cytomining/cytominer/blob/107101cb6670fb3ae519785e2e3c49afb44283d1/R/normalize.R) version, which also uses "standardize". The [February 23rd, 2017](https://github.com/broadinstitute/cytominer_scripts/blob/5e1342cb00ab4f62e704d9cf1689ec719f665384/normalize.R#L13) `cytominer_scripts/normalize` also uses robustize.
3. I think the current implementation of `cytominer_scripts/normalize`[ _samples_ profiles before normalization](https://github.com/broadinstitute/cytominer_scripts/blob/master/normalize.R#L99). (Maybe I am reading the code wrong though). Without knowing the seed used in the processing, there is no way to simulate this

**Comment by shntnu** [comment]
> * The file shows [normalization by DMSO control](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log.sh#L52). For some reason, I thought that the cytominer profiles were whole-plate normalized. This could be a reason for the discrepancy!!

My apologies for the confusion (in case I introduced it), but we did do a whole-plate normalization
when creating profiles for the "pseudo" batch `2016_04_01_a549_48hr_batch1_cmap_style`. 


[This](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/9e5832c6769760097a152b71449435262d2fcba1/analysis_log_cmap_style.sh#L79 ) was the script. There are two different sets of files on S3. 
- One corresponding to `2016_04_01_a549_48hr_batch1`, which was [normalization by DMSO control](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log.sh#L52)
- The other corresponding to `2016_04_01_a549_48hr_batch1_cmap_style`, which was  [normalization by the whole plate](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log_cmap_style.sh#L79), but everything else upstream of normalization was identical to `2016_04_01_a549_48hr_batch1`.

I think you are using `2016_04_01_a549_48hr_batch1`. An easy way to check is to compute the mean DMSO profile per plate; it should be all zeros (within some epsilon). Will do so.

For our notes: here are the two sets of files for a sample plate.
```
~$ aws s3 ls s3://imaging-platform-cold/imaging_analysis/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/plates/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad_2016_04_01_a549_48hr_batch1_SQ00014812_backend.tar.gz
2018-06-15 20:18:12 9778892027 2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad_2016_04_01_a549_48hr_batch1_SQ00014812_backend.tar.gz
~$ aws s3 ls s3://imaging-platform-cold/imaging_analysis/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/plates/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad_2016_04_01_a549_48hr_batch1_cmap_style_SQ00014812_backend.tar.gz
2018-06-19 16:05:52   31196865 2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad_2016_04_01_a549_48hr_batch1_cmap_style_SQ00014812_backend.tar.gz
```

**Comment by shntnu** [comment]
> * One corresponding to `2016_04_01_a549_48hr_batch1`, which was [robustize by DMSO control](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log.sh#L52)
> * The other corresponding to `2016_04_01_a549_48hr_batch1_cmap_style`, which was  [robustize by the whole plate](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log_cmap_style.sh#L79),

I can confirm this

See these two plots in the attached notebook
- `x2016_04_01_a549_48hr_batch1_cmap_style_SQ00015233_normalized_median_full_plate`
- `x2016_04_01_a549_48hr_batch1_SQ00015233_normalized_median_dmso`
[resolving_normalization_issue.nb.html.zip](https://github.com/broadinstitute/lincs-cell-painting/files/4611716/resolving_normalization_issue.nb.html.zip)

**Comment by gwaygenomics** [comment]
>>> This could be a reason for the discrepancy!!

>> * One corresponding to `2016_04_01_a549_48hr_batch1`, which was [robustize by DMSO control](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log.sh#L52)
>> * The other corresponding to `2016_04_01_a549_48hr_batch1_cmap_style`, which was  [robustize by the whole plate](https://github.com/broadinstitute/2015_10_05_DrugRepurposing_AravindSubramanian_GolubLab_Broad/blob/master/analysis_log_cmap_style.sh#L79),

> I can confirm this

Looks like we are on to something here 👇 (after removing those pesky 10 nonuniform plates)

![summary_metrics_full_uniformplates](https://user-images.githubusercontent.com/7340421/81598242-fbf80600-9394-11ea-8690-dbd7dbe08244.png)

![summary_metrics_zoom_uniformplates](https://user-images.githubusercontent.com/7340421/81598246-fe5a6000-9394-11ea-9789-2abb0d0629d4.png)

![level_4a_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81604701-4cc02c80-939e-11ea-9cbe-27af20d27332.png)

![level_4b_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81604907-a6c0f200-939e-11ea-86aa-c6c61cb201d8.png)

Check out those axes! 👀

**Comment by gwaygenomics** [comment]
I am much more comfortable with these comparisons 👍 a couple of takeaways:

* The feature selection comparisons are within floating point estimates, but there is some wobble. There is more wobble in the 4a profiles (before feature selection). I think this is why feature selection tends to work well in our cases. 
  * Our normalization procedures do not play nicely with all CellProfiler feature outputs.
  * This **will** change experiment-to-experiment, and might even change between CellProfiler versions.
  * CellProfiler features that are introduced with each analysis module could come alongside a metadata spreadsheet for data scientists that describes expected feature behavior and normalization strategies.
  * I bet that a _dynamic_ normalization scheme that automatically recognizes distributions of features and scales accordingly will give us (probably only moderate) performance boosts.
* I am happy with this comparison and think that we should move forward with the pycytominer profiles for a version 1 release. Like I mentioned previously, we can always rerelease an updated version!

One additional thing I should do before merging is to compare the pycytominer feature selected (4b) to the cytominer normalized (4a). This will tell us if the pycytominer feature selection is removing noisy features well enough (currently the 4b plots are based on intersections). I think this is actually the last step.

**Comment by shntnu** [comment]
> * I am happy with this comparison and think that we should move forward with the pycytominer profiles for a version 1 release. Like I mentioned previously, we can always rerelease an updated version!

Agree

**Comment by shntnu** [comment]
> * I bet that a _dynamic_ normalization scheme that automatically recognizes distributions of features and scales accordingly will give us (probably only moderate) performance boosts.

 distributions of features at the aggregate level or single cell level?

**Comment by gwaygenomics** [comment]
> distributions of features at the aggregate level or single cell level?

Aggregate level, but probably would work at single cell right? I haven't played with single cell yet so I don't know these distributions

**Comment by gwaygenomics** [comment]
> One additional thing I should do before merging is to compare the pycytominer feature selected (4b) to the cytominer normalized (4a). This will tell us if the pycytominer feature selection is removing noisy features well enough (currently the 4b plots are based on intersections). I think this is actually the last step.

## Comparing Cytominer 4a to Pyctyominer 4b

Results below. There are a couple discrepancies but not much more than we see in 4b intersection.

<details>
  <summary>Click to show figures</summary>

![pycytominer_select_metrics_per_feature](https://user-images.githubusercontent.com/7340421/81614083-8482a080-93ad-11ea-89e3-d7d65ab77cd4.png)

![pycytominer_select_metrics_per_plate](https://user-images.githubusercontent.com/7340421/81614094-877d9100-93ad-11ea-9875-54ccc4420396.png)
</details>

**Comment by gwaygenomics** [comment]
closed by merging #29
